### ISSUE TITLE
`GET /totp/exists` can have side-effects, which seems confusing

### ISSUE THREAD

**Issue by rfk** [mozilla/fxa-auth-server#2541]
Over in https://github.com/mozilla/PyFxA/pull/63, @kewisch was working on adding TOTP support to PyFxA but was seeing unexpected failures in the tests.  It turns out they were hitting this behaviour where `GET /totp/exists` will delete an unverified TOTP token from the account:

  https://github.com/mozilla/fxa-auth-server/blob/master/lib/routes/totp.js#L206

Where the comment says:

```
// If the token is not verified, lets delete it and report that
// it doesn't exist. This will help prevent some edge
// cases where the user started creating a token but never completed.
```

As a consumer of the `/totp/*` APIs, this seems surprising - it's a `GET` effect that can have side-effects that make a recently-added TOTP token disappear.  What are some of the edge-cases here, and is there a way for us to address them without this deletion-as-a-side-effect behaviour?

/cc @shane-tomlinson @vbudhram

**Comment by vbudhram** [comment]
@rfk Based on our initial [designs](https://www.lucidchart.com/documents/view/833ea3ae-65bc-4b0a-94cb-3290ec27ae85) we did not have intermediate states where the user might have cancelled while in the TOTP setup process. The assumption at that point was that an unverified token should be considered non-existing because it wouldn't be able to be verified via web content. The edge cases are when the user refreshes/navigates away from the setup flow. 

The `GET /totp/exists` endpoint tried to clean up the token because of this.

> is there a way for us to address them without this deletion-as-a-side-effect behavior?

Definitely, I could see it looking something like this

1. Content-server performs the `GET /totp/exists`, which also returns a verification flag
2. If it token is unverified, content-server should delete it, re-create token and show new QR code
3. If token is verified, show option to delete it

There is an extra request but I think that might be ok. FWIW, I'm 👍🏽 on making this change since it will make things less confusing.

**Comment by rfk** [comment]
Makes sense.  Another option could be to change the behaviour of `/exists` in conjunction with `/create`.  Right now this pair behave as follows:

* `/create` will error out if the user has a TOTP token, verified or otherwise
* `/exists` will delete an unverified TOTP token and return false

Which means that a sequence like the following is possible:

* `POST /totp/create`, which successfully creates an unverified token
* `POST /totp/create` again, which will error out because a token already exists
* `GET /totp/exists`, which returns `false` and deletes the unverified token
* `POST /totp/create` again, which will not succeed because the unverified token was deleted

We could change it so that `/exists` still reports `false` on an unverified token, but does not delete it.  Instead, we allow `/create` to overwrite an existing unverified token, so the above sequence goes like:

* `POST /totp/create`, which successfully creates an unverified token
* `POST /totp/create` again, which overwrites it with a new unverified token
* `GET /totp/exists`, which returns `false` but has no side-effects
* `POST /totp/create` again, which again overwrites it with a new unverified token
* `POST /session/verify/totp` which verifies the most-recently-created token
* `GET /totp/exists`, which now returns `true` due to the token being verified

Would this give us the behaviour you describe for the content-server in those edge-cases, without having to add more requests and branching logic into content-server?

**Comment by shane-tomlinson** [comment]
> Would this give us the behaviour you describe for the content-server in those edge-cases, without having to add more requests and branching logic into content-server?

The proposed behavior makes a lot of sense to me. 

@vbudhram - Is the intention of deleting unverified tokens in /totp/exists to cause unverified tokens to be deleted whenever the /settings page is loaded? If so, I'm wondering if there is another way to achieve the same effect, perhaps using a token expiration date.

**Comment by vbudhram** [comment]
> unverified tokens to be deleted whenever the /settings page is loaded

Correct, when page is reloaded that token can not be verified via the web anymore since we don't know what they did with QR. The old token was removed and a new one was created and shown.

> perhaps using a token expiration date

Possibly, I am not sure what a reasonable expiration date would be though.

@rfk Your proposal seems reasonable and would give us the intended behavior.
